File: python/src/aun_core/_cert_utils.py
from __future__ import annotations

import base64
import hashlib
import re
from datetime import datetime, timezone
from typing import Any

from cryptography import x509
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec, ed25519
from cryptography.x509.oid import NameOID

from .errors import AuthError


AGENT_MD_SIGNATURE_MARKER = "<!-- AUN-SIGNATURE"
_AGENT_MD_SIGNATURE_RE = re.compile(
    r"<!-- AUN-SIGNATURE\r?\n(?P<body>.*?)\r?\n-->\s*\Z",
    re.DOTALL,
)
_FINGERPRINT_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")


def normalize_fingerprint_hex(value: str | None) -> str:
    text = str(value or "").strip().lower()
    if text.startswith("sha256:"):
        text = text.split(":", 1)[1]
    text = text.replace(":", "")
    if len(text) not in (16, 64) or not _FINGERPRINT_HEX_RE.fullmatch(text):
        return ""
    return text


def cert_fingerprint_hexes(cert: x509.Certificate) -> tuple[str, str]:
    cert_hex = cert.fingerprint(hashes.SHA256()).hex()
    spki_hex = hashlib.sha256(public_key_der(cert.public_key())).hexdigest()
    return cert_hex, spki_hex


def cert_matches_fingerprint(cert: x509.Certificate, fingerprint: str | None) -> bool:
    expected = normalize_fingerprint_hex(fingerprint)
    if not expected:
        return False
    cert_hex, spki_hex = cert_fingerprint_hexes(cert)
    if len(expected) == 16:
        return cert_hex[:16] == expected or spki_hex[:16] == expected
    return cert_hex == expected or spki_hex == expected


def public_key_matches_fingerprint(cert: x509.Certificate, fingerprint: str | None) -> bool:
    expected = normalize_fingerprint_hex(fingerprint)
    if not expected:
        return False
    _, spki_hex = cert_fingerprint_hexes(cert)
    if len(expected) == 16:
        return spki_hex[:16] == expected
    return spki_hex == expected


def parse_agent_md_tail_signature(content: str) -> tuple[str, dict[str, str] | None, str | None]:
    marker_index = content.rfind(AGENT_MD_SIGNATURE_MARKER)
    if marker_index < 0:
        return content, None, None
    if marker_index > 0 and content[marker_index - 1] not in "\r\n":
        return content, None, None

    tail = content[marker_index:]
    match = _AGENT_MD_SIGNATURE_RE.fullmatch(tail)
    if not match:
        return content[:marker_index], None, "malformed signature block"

    fields: dict[str, str] = {}
    for line in match.group("body").splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        if ":" not in stripped:
            return content[:marker_index], None, f"malformed signature field: {stripped}"
        key, value = stripped.split(":", 1)
        fields[key.strip().lower()] = value.strip()

    for required in ("cert_fingerprint", "timestamp", "signature"):
        if not fields.get(required):
            return content[:marker_index], None, f"signature block missing {required}"
    if not normalize_fingerprint_hex(fields["cert_fingerprint"]):
        return content[:marker_index], None, "invalid cert_fingerprint"
    if fields.get("public_key_fingerprint") and not normalize_fingerprint_hex(fields["public_key_fingerprint"]):
        return content[:marker_index], None, "invalid public_key_fingerprint"
    try:
        int(fields["timestamp"])
    except ValueError:
        return content[:marker_index], None, "invalid timestamp"

    return content[:marker_index], fields, None


def extract_agent_md_aid(payload: str) -> str:
    lines = payload.lstrip("\ufeff").splitlines()
    if not lines or lines[0].strip() != "---":
        return ""
    for line in lines[1:]:
        stripped = line.strip()
        if stripped == "---":
            break
        if stripped.startswith("aid:"):
            value = stripped.split(":", 1)[1].strip()
            if len(value) >= 2 and value[0] == value[-1] and value[0] in {"'", '"'}:
                value = value[1:-1]
            return value.strip()
    return ""


def build_agent_md_signature_block(
    *,
    cert_fingerprint: str,
    timestamp: int,
    signature_b64: str,
    public_key_fingerprint: str = "",
) -> str:
    lines = [
        "<!-- AUN-SIGNATURE",
        f"cert_fingerprint: {cert_fingerprint}",
    ]
    if public_key_fingerprint:
        lines.append(f"public_key_fingerprint: {public_key_fingerprint}")
    lines.extend([
        f"timestamp: {int(timestamp)}",
        f"signature: {signature_b64}",
        "-->",
    ])
    return "\n".join(lines)


def normalize_agent_md_payload(content: str) -> str:
    payload, _, _ = parse_agent_md_tail_signature(str(content or ""))
    if payload and not payload.endswith(("\n", "\r")):
        payload += "\n"
    return payload


def verify_signature(public_key: Any, sig_bytes: bytes, data_bytes: bytes) -> None:
    if isinstance(public_key, ed25519.Ed25519PublicKey):
        public_key.verify(sig_bytes, data_bytes)
        return
    if isinstance(public_key, ec.EllipticCurvePublicKey):
        if isinstance(public_key.curve, ec.SECP384R1):
            public_key.verify(sig_bytes, data_bytes, ec.ECDSA(hashes.SHA384()))
        else:
            public_key.verify(sig_bytes, data_bytes, ec.ECDSA(hashes.SHA256()))
        return
    raise AuthError(f"unsupported identity public key type: {type(public_key)!r}")


def sign_bytes(private_key: Any, payload: bytes) -> bytes:
    if isinstance(private_key, ed25519.Ed25519PrivateKey):
        return private_key.sign(payload)
    if isinstance(private_key, ec.EllipticCurvePrivateKey):
        if isinstance(private_key.curve, ec.SECP384R1):
            return private_key.sign(payload, ec.ECDSA(hashes.SHA384()))
        return private_key.sign(payload, ec.ECDSA(hashes.SHA256()))
    raise AuthError(f"unsupported identity private key type: {type(private_key)!r}")


def verify_bytes(public_key: Any, payload: bytes, signature_b64: str) -> bool:
    try:
        signature = base64.b64decode(str(signature_b64 or ""), validate=True)
        verify_signature(public_key, signature, payload)
        return True
    except InvalidSignature:
        return False


def cert_common_name(cert: x509.Certificate, *, issuer: bool = False) -> str:
    name = cert.issuer if issuer else cert.subject
    attrs = name.get_attributes_for_oid(NameOID.COMMON_NAME)
    return str(attrs[0].value) if attrs else ""


def cert_fingerprint(cert: x509.Certificate) -> str:
    return "sha256:" + cert.fingerprint(hashes.SHA256()).hex()


def public_key_fingerprint(cert: x509.Certificate) -> str:
    return "sha256:" + hashlib.sha256(public_key_der(cert.public_key())).hexdigest()


def public_key_der_b64(cert: x509.Certificate) -> str:
    der = cert.public_key().public_bytes(
        serialization.Encoding.DER,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return base64.b64encode(der).decode("ascii")


def public_key_der(public_key: Any) -> bytes:
    return public_key.public_bytes(
        serialization.Encoding.DER,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )


def payload_bytes(payload: bytes | str) -> bytes:
    if isinstance(payload, bytes):
        return payload
    return str(payload).encode("utf-8")


def cert_time_error(cert: x509.Certificate) -> str:
    now = datetime.now(timezone.utc)
    if now < cert.not_valid_before_utc:
        return "not_yet_valid"
    if now > cert.not_valid_after_utc:
        return "expired"
    return ""

File: ts/src/cert-utils.ts
/**
 * 证书与签名工具函数 — 对齐 Python SDK _cert_utils.py
 * 从 namespaces/auth.ts 提取的纯函数，供 AID/AIDStore 使用。
 */

import { createSign, createVerify, X509Certificate, createHash } from 'node:crypto';
import { certificateSha256Fingerprint } from './crypto.js';

const AGENT_MD_SIGNATURE_MARKER = '<!-- AUN-SIGNATURE';
const AGENT_MD_SIGNATURE_RE = /^<!-- AUN-SIGNATURE\r?\n(?<body>[\s\S]*?)\r?\n-->\s*$/;
const AGENT_MD_FINGERPRINT_RE = /^sha256:[0-9a-f]{64}$/;

/** 解析 agent.md 尾部签名块 */
export function parseAgentMdTailSignature(content: string): {
  payload: string;
  fields: Record<string, string> | null;
  parseError?: string;
} {
  const idx = content.lastIndexOf(AGENT_MD_SIGNATURE_MARKER);
  if (idx < 0) return { payload: content, fields: null };
  if (idx > 0 && content[idx - 1] !== '\n' && content[idx - 1] !== '\r') {
    return { payload: content, fields: null };
  }
  const tail = content.slice(idx);
  const match = tail.match(AGENT_MD_SIGNATURE_RE);
  if (!match) {
    return { payload: content.slice(0, idx), fields: null, parseError: 'malformed signature block' };
  }
  const fields: Record<string, string> = {};
  for (const rawLine of match.groups?.body?.split(/\r?\n/) ?? []) {
    const line = rawLine.trim();
    if (!line) continue;
    const colon = line.indexOf(':');
    if (colon < 0) {
      return { payload: content.slice(0, idx), fields: null, parseError: `malformed signature field: ${line}` };
    }
    fields[line.slice(0, colon).trim().toLowerCase()] = line.slice(colon + 1).trim();
  }
  for (const req of ['cert_fingerprint', 'timestamp', 'signature']) {
    if (!fields[req]) {
      return { payload: content.slice(0, idx), fields: null, parseError: `signature block missing ${req}` };
    }
  }
  if (!AGENT_MD_FINGERPRINT_RE.test(fields.cert_fingerprint.toLowerCase())) {
    return { payload: content.slice(0, idx), fields: null, parseError: 'invalid cert_fingerprint' };
  }
  if (!Number.isFinite(Number(fields.timestamp))) {
    return { payload: content.slice(0, idx), fields: null, parseError: 'invalid timestamp' };
  }
  return { payload: content.slice(0, idx), fields };
}

/** 从 agent.md frontmatter 提取 aid 字段 */
export function extractAgentMdAid(payload: string): string {
  const lines = payload.replace(/^﻿/, '').split(/\r?\n/);
  if (!lines.length || lines[0].trim() !== '---') return '';
  for (const line of lines.slice(1)) {
    const t = line.trim();
    if (t === '---') break;
    if (t.startsWith('aid:')) {
      let v = t.slice(4).trim();
      if (v.length >= 2 && v[0] === v[v.length - 1] && (v[0] === '"' || v[0] === "'")) {
        v = v.slice(1, -1);
      }
      return v.trim();
    }
  }
  return '';
}

/** 规范化 agent.md payload（去除签名块，确保末尾换行） */
export function normalizeAgentMdPayload(content: string): string {
  let payload = parseAgentMdTailSignature(String(content ?? '')).payload;
  if (payload && !payload.endsWith('\n') && !payload.endsWith('\r')) payload += '\n';
  return payload;
}

/** 构造 agent.md 签名块 */
export function buildAgentMdSignatureBlock(certFingerprint: string, timestamp: number, signatureB64: string): string {
  return [
    '<!-- AUN-SIGNATURE',
    `cert_fingerprint: ${certFingerprint}`,
    `timestamp: ${Math.trunc(timestamp)}`,
    `signature: ${signatureB64}`,
    '-->',
  ].join('\n');
}

/** 使用私钥签名（ECDSA P-256 SHA-256，DER 编码输出） */
export function signBytes(privateKeyPem: string, payload: Buffer): Buffer {
  const signer = createSign('SHA256');
  signer.update(payload);
  signer.end();
  return signer.sign(privateKeyPem);
}

/** 使用证书公钥验签 */
export function verifySignatureWithCert(certPem: string, signature: Buffer, data: Buffer): boolean {
  try {
    const cert = new X509Certificate(certPem);
    const verifier = createVerify('SHA256');
    verifier.update(data);
    verifier.end();
    return verifier.verify(cert.publicKey, signature);
  } catch {
    return false;
  }
}

/** 获取证书 Subject CN */
export function certCommonName(certPem: string, issuer = false): string {
  try {
    const cert = new X509Certificate(certPem);
    const dn = issuer ? cert.issuer : cert.subject;
    const match = dn.match(/(?:^|,\s*)CN=([^,\n]+)/);
    return match?.[1]?.trim() ?? '';
  } catch {
    return '';
  }
}

/** 检查证书有效期，返回 '' | 'expired' | 'not_yet_valid' */
export function certTimeError(certPem: string): '' | 'expired' | 'not_yet_valid' {
  try {
    const cert = new X509Certificate(certPem);
    const now = Date.now();
    if (now < new Date(cert.validFrom).getTime()) return 'not_yet_valid';
    if (now > new Date(cert.validTo).getTime()) return 'expired';
    return '';
  } catch {
    return 'expired';
  }
}

/** 获取证书公钥 DER base64 */
export function publicKeyDerB64(certPem: string): string {
  try {
    const cert = new X509Certificate(certPem);
    return cert.publicKey.export({ type: 'spki', format: 'der' }).toString('base64');
  } catch {
    return '';
  }
}

/** 证书 SHA-256 指纹（复用 crypto.ts 的实现） */
export { certificateSha256Fingerprint as certFingerprint };
